### DISCUSSION TITLE
CTC beam search without language model

### DISCUSSION THREAD

**Discussion by hamjam** [NVIDIA-NeMo/NeMo#4064] [Q&A]
Hi and thank you for your great toolkit,

@titu1994 
It seems that there is not any CTC decoding strategy for CTC models except using `BeamSearchDecoderWithLM` class or greedy strategy.  First, there is no mention in the `BeamSearchDecoderWithLM`'s docstring for how to perform beam search without lm. Then, the class just uses `ctc_decoders` from OpenSeq2Seq and it doesn't implement beam search directly for more flexibility. Is there any limitation to mentioned problems? Is it possible to implement CTC beam search without language modeling?

Thanks in advance

**Comment by VahidooX** [comment]
ctc_decoders is a very fast implementation of beam search decoding. You can use them with or without LM. You may just pass None as the lm_path to have a regular beam search decoding. You may take a look here in the docs and the scripts to learn more on the detail:

https://github.com/NVIDIA/NeMo/blob/main/scripts/asr_language_modeling/ngram_lm/eval_beamsearch_ngram.py

https://docs.nvidia.com/deeplearning/nemo/user-guide/docs/en/main/asr/asr_language_modeling.html

If you are looking to make changes to the beam search decoding, then you can use this library which is fully pythonic and works with nemo models:

https://github.com/kensho-technologies/pyctcdecode

https://blog.kensho.com/pyctcdecode-a-new-beam-search-decoder-for-ctc-speech-recognition-2be3863afa96
